pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,233 @@
+	cbMetrics := &circuitBreakerMetrics{
+		circuitBreakerTransitions: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingester_circuit_breaker_transitions_total",
+			Help: "Number of times the circuit breaker has entered a state.",
+		}, []string{"state"}),
+		circuitBreakerResults: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingester_circuit_breaker_results_total",
+			Help: "Results of executing requests via the circuit breaker.",
+		}, []string{"result"}),
+	}
+	circuitBreakerCurrentStateGaugeFn := func(state circuitbreaker.State) prometheus.GaugeFunc {
+		return promauto.With(r).NewGaugeFunc(prometheus.GaugeOpts{
+			Name:        "cortex_ingester_circuit_breaker_current_state",
+			Help:        "Boolean set to 1 whenever the circuit breaker is in a state corresponding to the label name.",
+			ConstLabels: map[string]string{"state": state.String()},
+		}, func() float64 {
+			if currentStateFn() == state {
+				return 1
+			}
+			return 0
+		})
+	}
+	for _, s := range []circuitbreaker.State{circuitbreaker.OpenState, circuitbreaker.HalfOpenState, circuitbreaker.ClosedState} {
+		circuitBreakerCurrentStateGaugeFn(s)
+		// We initialize all possible states for the circuitBreakerTransitions metrics
+		cbMetrics.circuitBreakerTransitions.WithLabelValues(s.String())
+	}
+	for _, r := range []string{resultSuccess, resultError, resultOpen} {
+		// We initialize all possible results for the circuitBreakerResults metrics
+		cbMetrics.circuitBreakerResults.WithLabelValues(r)
+	}
+	return cbMetrics
+}
+
+type CircuitBreakerConfig struct {
+	Enabled                    bool          `yaml:"enabled" category:"experimental"`
+	FailureThresholdPercentage uint          `yaml:"failure_threshold_percentage" category:"experimental"`
+	FailureExecutionThreshold  uint          `yaml:"failure_execution_threshold" category:"experimental"`
+	ThresholdingPeriod         time.Duration `yaml:"thresholding_period" category:"experimental"`
+	CooldownPeriod             time.Duration `yaml:"cooldown_period" category:"experimental"`
+	InitialDelay               time.Duration `yaml:"initial_delay" category:"experimental"`
+	PushTimeout                time.Duration `yaml:"push_timeout" category:"experiment"`
+	testModeEnabled            bool          `yaml:"-"`
+}
+
+func (cfg *CircuitBreakerConfig) RegisterFlags(f *flag.FlagSet) {
+	prefix := "ingester.circuit-breaker."
+	f.BoolVar(&cfg.Enabled, prefix+"enabled", false, "Enable circuit breaking when making requests to ingesters")
+	f.UintVar(&cfg.FailureThresholdPercentage, prefix+"failure-threshold-percentage", 10, "Max percentage of requests that can fail over period before the circuit breaker opens")
+	f.UintVar(&cfg.FailureExecutionThreshold, prefix+"failure-execution-threshold", 100, "How many requests must have been executed in period for the circuit breaker to be eligible to open for the rate of failures")
```

Comment:
Can we rename this to something like "minimum requests count in period for circuit-breaker to open"?

Reply:
I’ve used the same config as in failsafe-go, although so etimes they were not the best ones. I think that if we keep the configs consistent, it us easier to follow both imementations (Mimir circuit breakers and fail-safe circuit breakers). So if you are not strongly against this I’d keep it like this.